virtio-devices: Add rate limiter for the virtio-net device

likebreath opened the pull request:
This PR add the rate limiter for virtio-net devices. Also, to resolve a dependency issue, the `rate_limiter` module is now moved into its own crate  (from crate `virtio-devices`). Note: the RX and TX queue share the same `rate limit` from user inputs, but they are throttled independently.

Fixes: #1286

Signed-off-by: Bo Chen <chen.bo@intel.com>

likebreath commented:
While my experiments locally (similar to what @sboeuf did in #1731 w/ iperf3) showed the rate limiting is working fine. But later I noticed the current PR will regularly get the following error message while running iperf3:

```
cloud-hypervisor: 36.035457154s: <_net1_qp0> ERROR:virtio-devices/src/net.rs:110 -- Failed to get tx queue event: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
```

I will take a closer look tomorrow, likely related to how I should restart processing the queue after cool-down of rate-limiting correctly.

likebreath commented:
I did the following experiments on the rate-limiting with this PR. The experiment setup is borrowed from @sboeuf's issue #1731.

#### Cloud hypervisor command-line
```
./cloud-hypervisor/target/release/cloud-hypervisor \
    --api-socket /tmp/ch-socket \
    --cpus boot=4 \
    --memory size=4G \
    --kernel bzImage \
    --cmdline "root=/dev/vda1 console=hvc0 rw" \
    --disk path=focal-server-cloudimg-amd64-custom-20210106-1.raw
```

#### iperf server command-line
```
iperf3 -s -p 5201 &
```

#### Hotplug a virtio-net with rate limitting of `800Mbits/sec` (e.g. `10M/100ms`)
```
./cloud-hypervisor/target/release/ch-remote \
    --api-socket /tmp/ch-socket \
    add-net \
    tap=,tx_bw_size=10485760,tx_bw_refill_time=100
```

#### Guest commands:

```
sudo ip addr add 192.168.249.2/24 dev ens4
sudo ip link set up dev ens4

iperf3 -c 192.168.249.1 -p 5201 -t 60s
```

The output of iperf3 confirms the rate limiting is in place, with the following output:

```
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-60.00  sec  5.83 GBytes   835 Mbits/sec    0             sender
[  5]   0.00-60.00  sec  5.83 GBytes   835 Mbits/sec                  receiver

iperf Done.
```

#### Additional notes
Note that the reported Bitrate from iperf3 (853 Mbits/sec) does not match perfectly with the rate limiter from user, but I think the differences are expected and acceptable (considering the rate-limiter is based on TokenBucket and also actual Bitrate is workload specific). As a reference, Firecracker allows to have 10% differences for their rate-limiter [integration tests](https://github.com/firecracker-microvm/firecracker/blob/master/tests/integration_tests/functional/test_rate_limiter.py#L29).

What's more, I did a similar experiment with a multi-queue setup (2 queue-pairs), and confirmed the rate-limiting is working fine. I also measured the performance of virtio-net with this PR when no rate-limiter is used, and compared it with the performance data from the upstream branch. With single queue-pair setup, I got 23.9 Gbits/sec (for the PR branch) 24.2 Gbits/sec (for the upstream). Similarly, with 2 queue-pairs, I got (14.4+14.3) Gbits/sec and (14.8 + 14.9) Gbits/sec. In general, I found this PR does not bring noticeable overhead to virtio-net performance.

sboeuf commented on virtio-devices/src/block.rs:
```diff
@@ -41,6 +40,7 @@ use vm_migration::{
     Migratable, MigratableError, Pausable, Snapshot, SnapshotDataSection, Snapshottable,
     Transportable,
 };
+use vm_virtio::rate_limiter::{RateLimiter, TokenType};
```
Since the rate limiter is not virtio specific at all, and could easily be reused for non virtio devices, I think it makes sense to move it to its own crate. I'm also thinking we shouldn't add more things to `vm-virtio` as it will make the move to the upstream version harder as well.
WDYT @rbradford ?

rbradford replied:
Definitely not in the vm-virtio crate. Leave in the virtio-devices crate for now and can move if we need it elsewhere (don't see anything on the immediate horizon though).

sboeuf replied:
No he can't leave it in `virtio-devices` because there's some cyclic dependencies otherwise.

rbradford replied:
Okey dokey. New crate it is then. :+1:

likebreath replied:
Got it. Will move it to its own crate.

sboeuf reviewed (commented):
Looks good to me, except the pending question about where to move the `rate_limiter`.

rbradford commented:
Since this is release week I don't want to merge the TX side without the RX side also handled.

likebreath commented:
> Since this is release week I don't want to merge the TX side without the RX side also handled.

Sure. I am working on the RX side, and will reuse this PR for that. Let me try to get a working version today.

likebreath commented:
@sboeuf @rbradford Please find the updated PR and experiments results above. 

The only open I have now is on actual RX rate reported by iperf3 (e.g. `527 Mbits/sec`) which is quite different from the rate limit from user input (`800 Mbits/sec`). I tried to disable the RX_TAP_EVEN (by `unregister_listener` and setting `rx_tap_listening=false`) while the `rx_rate_limiter` is blocking, but it seems to be not helpful (debug branch here: https://github.com/likebreath/cloud-hypervisor/commit/0bd1a804f352ba640e89d3c864c7ac4717b0f634).

likebreath commented:
@sboeuf @rbradford PTAL. The PR is updated and ready for review.  The experiments data from https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2401#issuecomment-802383132 is also updated.

sboeuf commented on virtio-devices/src/net.rs:
```diff
@@ -90,7 +92,14 @@ impl NetEpollHandler {
+            .rx_rate_limiter
+            .as_ref()
+            .map_or(false, |r| r.is_blocked());
+
```
In here, you should set the new boolean that will store the information about some RX buffers being ready. Even if the limit is reached at the time the RX buffer is added, it's important to have this information as it will help decide if the TAP fd should be registered or not when the `RX_RATE_LIMITER_EVENT` is triggered.

sboeuf commented on net_util/src/queue_pair.rs:
```diff
@@ -193,6 +241,30 @@ impl RxVirtio {
+                if rate_limit_reached {
+                    // Set the deferred frame which will be processed when
+                    // the 'rx_rate_limiter' is unblocked
+                    *rx_deferred_frame = Some(len);
```
I don't understand the need for introducing a `rx_deferred_frame` and the whole logic that comes with it.
I was expecting that if we reached the rate limiter limit, then we go to the previous descriptor and we're done with it. It's important to unregister the TAP fd after we did so, but nothing else.
Is there an aspect that I'm missing and that's mandatory to make the rate limiter work? I'm asking because this adds a lot of complexity to the code.

likebreath replied:
We need the `rx_deferred_frame` as we can't revert the read from the TAP when the rate limit is reached. Note that we are only able to check whether a RX frame is over the rate limit after the `readv` returns the actual `bytes` being read from the TAP.

rbradford commented on rate_limiter/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "rate_limiter"
+version = "0.1.0"
+authors = ["Amazon Firecracker team <firecracker-devel@amazon.com>"]
```
I think we should change it simply because I don't think the FC team want emails about some crate that we've created. Is `authors` even a mandatory field? i.e. could we simply remove it.

likebreath replied:
Cool. The `authors` filed is not needed unless we are publishing it to `crate.io`. Let's simply remove it now.

likebreath commented:
@sboeuf @rbradford PTAL. The PR is updated with the comments from our online discussion today: removed the `rx_deferred_frame` for the sake of simplicity. As measured, the current actual rate limit is very similar to what we got from the PR with the `rx_deferred_frame`: 840 Mbits/second vs 835 Mbits/second (with 800 Mbits/second limit from user).  The experiments result from https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2401#issuecomment-802383132 above is also updated. Any more comments or advises, please let me know. Thank you!

Note: the Clippy CI failure is being fixed from https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2435.

sboeuf commented on virtio-devices/src/net.rs:
```diff
@@ -209,6 +223,35 @@ impl EpollHelperHandler for NetEpollHandler {
+                        }
+                    }
+                } else {
+                    error!("Spurious RX_RATE_LIMITER_EVENT");
```
Nit: I think `Unexpected` would be more appropriate than `Spurious` here.

sboeuf commented on virtio-devices/src/net.rs:
```diff
@@ -209,6 +223,35 @@ impl EpollHelperHandler for NetEpollHandler {
+            RX_RATE_LIMITER_EVENT => {
+                if let Some(rate_limiter) = &mut self.net.rx_rate_limiter {
+                    // Upon rate limiter event, call the rate limiter handler
+                    // and restart processing the RX_TAP_EVENT
```
Nit: s/and restart processing the RX_TAP_EVENT/and register the TAP fd for further processing if some RX buffers are available/

sboeuf reviewed (approved):
Great PR! Two small nits, but otherwise it's ready to be merged IMO.
Also, let's wait for the clippy issues to be resolved and 0.14 to be released before we merge this PR.
